README.md
[![GitHub release](https://img.shields.io/github/release/cybozu-go/usocksd.svg?maxAge=60)][releases]
[![GoDoc](https://godoc.org/github.com/cybozu-go/usocksd?status.svg)][godoc]
[![main](https://github.com/cybozu-go/usocksd/actions/workflows/main.yml/badge.svg)](https://github.com/cybozu-go/usocksd/actions/workflows/main.yml)
[![Go Report Card](https://goreportcard.com/badge/github.com/cybozu-go/usocksd)](https://goreportcard.com/report/github.com/cybozu-go/usocksd)
[![License](https://img.shields.io/github/license/cybozu-go/usocksd.svg?maxAge=2592000)](LICENSE)

Micro SOCKS server
==================

**usocksd** is a SOCKS server written in Go.

[`usocksd/socks`](https://godoc.org/github.com/cybozu-go/usocksd/socks)
is a general purpose SOCKS server library.  usocksd is built on it.

Features
--------

* Support for SOCKS4, SOCKS4a, SOCK5

    * Only CONNECT is supported (BIND and UDP associate is missing).

* Graceful stop & restart

    * On SIGINT/SIGTERM, usocksd stops gracefully.
    * On SIGHUP, usocksd restarts gracefully.

* Access log

    Thanks to [`cybozu-go/log`](https://github.com/cybozu-go/log),
    usocksd can output access logs in structured formats including
    JSON.

* Specific network interface

    usocksd can be configured to use specific network interface
    for outgoing connections.

    It is extremely useful if you want to send all traffic to VPN/Wireguard device
    or you have multiple network cards.

* Multiple external IP addresses

    usocksd can be configured to use multiple external IP addresses
    for outgoing connections.

    usocksd keeps using the same external IP address for a client
    as much as possible.  This means usocksd can proxy passive FTP
    connections reliably.

    Moreover, you can use a [DNSBL][] service to exclude dynamically
    from using some undesirable external IP addresses.

* White- and black- list of sites

    usocksd can be configured to grant access to the sites listed
    in a white list, and/or to deny access to the sites listed in a
    black list.

    usocksd can block connections to specific TCP ports, too.

Install
-------

Use a recent version of Go.

```
go get -u github.com/cybozu-go/usocksd/...
```

Usage
-----

`usocksd [-h] [-f CONFIG]`

The default configuration file path is `/etc/usocksd.toml`.

In addition, `usocksd` implements [the common spec](https://github.com/cybozu-go/well#specifications) from [`cybozu-go/well`](https://github.com/cybozu-go/well).

usocksd does not have *daemon* mode.  Use systemd to run it on your background.

Configuration file format
-------------------------

`usocksd.toml` is a [TOML][] file.
All fields are optional.

```
[log]
filename = "/path/to/file"         # default to stderr.
level = "info"                     # critical, error, warning, info, debug
format = "plain"                   # plain, logfmt, json

[incoming]
port = 1080
metrics_port = 1081                # Port number to serve metrics
addresses = ["127.0.0.1"]          # List of listening IP addresses
allow_from = ["10.0.0.0/8"]        # CIDR network or IP address

[outgoing]
allow_sites = [                    # List of FQDN to be granted.
    "www.amazon.com",              # exact match
    ".google.com",                 # subdomain match
]
deny_sites = [                     # List of FQDN to be denied.
    ".2ch.net",                    # subdomain match
    "bad.google.com",              # deny a domain of *.google.com
    "",                            # "" matches non-FQDN (IP) requests.
]
deny_ports = [22, 25]              # Black list of outbound ports
iface = tun0                       # Outgoing traffic binds to specific network interface
addresses = ["12.34.56.78"]        # List of source IP addresses
dnsbl_domain = "some.dnsbl.org"    # to exclude black listed IP addresses
```

Tuning
------

If you see usocksd consumes too much CPU, try setting [`GOGC`][GOGC] to higher value, say **300**.

License
-------

[MIT](https://opensource.org/licenses/MIT)

[releases]: https://github.com/cybozu-go/usocksd/releases
[DNSBL]: https://en.wikipedia.org/wiki/DNSBL
[TOML]: https://github.com/toml-lang/toml
[godoc]: https://godoc.org/github.com/cybozu-go/usocksd
[GOGC]: https://golang.org/pkg/runtime/#pkg-overview


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## [Unreleased]

## [1.3.0] - 2023-03-30
### Added
- Provide some metrics in OpenMetrics format (#14).

### Changed
- Update dependencies
- Use GitHub Actions instead of Circle CI

## [1.2.0] - 2020-10-29
### Added
- SilenceLogs option to socks.Server by @AkihiroSuda (#8)

### Changed
- Update dependencies

## [1.1.0] - 2018-11-18
### Changed
- Handle renaming of cybozu-go/cmd to [cybozu-go/well][well]
- Introduce support for Go modules

## [1.0.0] - 2016-09-01
### Added
- usocksd now has its own SOCKS implementation under `socks` directory.
- Support for SOCKS4/4a thanks to the new implementation.
- usocksd now adopts [github.com/cybozu-go/cmd][cmd] framework.  
  As a result, it implements [the common spec][spec] including graceful restart.

### Changed
- The default configuration file path is now `/etc/usocksd.toml`.
- `log.file` config key is renamed to `log.filename`.

[well]: https://github.com/cybozu-go/well
[cmd]: https://github.com/cybozu-go/cmd
[spec]: https://github.com/cybozu-go/cmd/blob/master/README.md#specifications
[Unreleased]: https://github.com/cybozu-go/usocksd/compare/v1.3.0...HEAD
[1.3.0]: https://github.com/cybozu-go/usocksd/compare/v1.2.0...v1.3.0
[1.2.0]: https://github.com/cybozu-go/usocksd/compare/v1.1.0...v1.2.0
[1.1.0]: https://github.com/cybozu-go/usocksd/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/cybozu-go/usocksd/compare/v0.1...v1.0.0


config.go
package usocksd

import (
	"errors"
	"net"
	"strings"

	"github.com/BurntSushi/toml"
	"github.com/cybozu-go/well"
)

const (
	defaultPort        = 1080
	defaultMetricsPort = 1081
)

// IncomingConfig is a set of configurations to accept clients.
type IncomingConfig struct {
	Port         int
	MetricsPort  int `toml:"metrics_port"`
	Addresses    []net.IP
	AllowFrom    []string `toml:"allow_from"`
	allowSubnets []*net.IPNet
}

// OutgoingConfig is a set of configurations to connect to destinations.
type OutgoingConfig struct {
	AllowSites  []string `toml:"allow_sites"`
	DenySites   []string `toml:"deny_sites"`
	DenyPorts   []int    `toml:"deny_ports"`
	IFace       string   `toml:"iface"`
	Addresses   []net.IP
	DNSBLDomain string `toml:"dnsbl_domain"`
}

// Config is a struct tagged for TOML for usocksd.
type Config struct {
	Log      well.LogConfig `toml:"log"`
	Incoming IncomingConfig `toml:"incoming"`
	Outgoing OutgoingConfig `toml:"outgoing"`
}

// NewConfig creates and initializes Config.
func NewConfig() *Config {
	c := new(Config)
	c.Incoming.Port = defaultPort
	c.Incoming.MetricsPort = defaultMetricsPort
	return c
}

// Load loads a TOML file from path.
func (c *Config) Load(path string) error {
	md, err := toml.DecodeFile(path, c)
	if err != nil {
		return err
	}
	if len(md.Undecoded()) > 0 {
		return errors.New("Unknown config keys in " + path)
	}

	if len(c.Incoming.AllowFrom) > 0 {
		subnets := make([]*net.IPNet, 0, len(c.Incoming.AllowFrom))
		for _, s := range c.Incoming.AllowFrom {
			if strings.IndexByte(s, '/') == -1 {
				s = s + "/32"
			}
			_, n, err := net.ParseCIDR(s)
			if err != nil {
				return errors.New("Invalid network or IP address: " + s)
			}
			subnets = append(subnets, n)
		}
		c.Incoming.allowSubnets = subnets
	}

	c.Outgoing.AllowSites = toLowerStrings(c.Outgoing.AllowSites)
	c.Outgoing.DenySites = toLowerStrings(c.Outgoing.DenySites)

	return nil
}

func toLowerStrings(l []string) (nl []string) {
	for _, s := range l {
		nl = append(nl, strings.ToLower(s))
	}
	return
}

// allowIP tests if ip is allowed to connect to usocksd.
func (c *Config) allowIP(ip net.IP) bool {
	if len(c.Incoming.allowSubnets) == 0 {
		return true
	}
	for _, n := range c.Incoming.allowSubnets {
		if n.Contains(ip) {
			return true
		}
	}
	return false
}

func siteMatch(site, match string) bool {
	if len(match) > 0 && match[0] == '.' {
		return strings.HasSuffix(site, match)
	}
	return site == match
}

// allowFQDN tests if FQDN is granted to access or not.
func (c *Config) allowFQDN(fqdn string) bool {
	fqdn = strings.ToLower(fqdn)
	if len(c.Outgoing.AllowSites) > 0 {
		for _, match := range c.Outgoing.AllowSites {
			if siteMatch(fqdn, match) {
				goto CHECK_DENY
			}
		}
		return false
	}

CHECK_DENY:
	for _, match := range c.Outgoing.DenySites {
		if siteMatch(fqdn, match) {
			return false
		}
	}
	return true
}

// allowPort tests if port is legitimate for destination.
func (c *Config) allowPort(port int) bool {
	for _, p := range c.Outgoing.DenyPorts {
		if p == port {
			return false
		}
	}
	return true
}
